From d0276a97d8f68ad630b0256726840ac57d9679bc Mon Sep 17 00:00:00 2001
From: Omar Sandoval <osandov@osandov.com>
Date: Sun, 10 Jan 2016 01:50:22 -0800
Subject: [PATCH 1/4] Set class name on status bar

This is useful for, e.g., compton.
---
 dwm.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dwm.c b/dwm.c
index 9585683327c3..764e7581e519 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1807,6 +1807,9 @@ updatebars(void)
 		.background_pixmap = ParentRelative,
 		.event_mask = ButtonPressMask|ExposureMask
 	};
+	XClassHint *ch = XAllocClassHint();
+	ch->res_name = "dwmstatus";
+	ch->res_class = "dwm";
 	for (m = mons; m; m = m->next) {
 		if (m->barwin)
 			continue;
@@ -1815,7 +1818,9 @@ updatebars(void)
 		                          CWOverrideRedirect|CWBackPixmap|CWEventMask, &wa);
 		XDefineCursor(dpy, m->barwin, cursor[CurNormal]->cursor);
 		XMapRaised(dpy, m->barwin);
+		XSetClassHint(dpy, m->barwin, ch);
 	}
+	XFree(ch);
 }
 
 void
-- 
2.9.0

