From 01156d95874d801b500948f25cbbd3a0ea328955 Mon Sep 17 00:00:00 2001
From: Omar Sandoval <osandov@osandov.com>
Date: Sun, 10 Jan 2016 01:50:22 -0800
Subject: [PATCH 1/2] Set class name on status bar

This is useful for, e.g., compton.
---
 dwm.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dwm.c b/dwm.c
index 4782343..f2ce061 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1809,6 +1809,9 @@ updatebars(void)
 		.background_pixmap = ParentRelative,
 		.event_mask = ButtonPressMask|ExposureMask
 	};
+	XClassHint *ch = XAllocClassHint();
+	ch->res_name = "dwmstatus";
+	ch->res_class = "dwm";
 	for (m = mons; m; m = m->next) {
 		if (m->barwin)
 			continue;
@@ -1817,7 +1820,9 @@ updatebars(void)
 				CWOverrideRedirect|CWBackPixmap|CWEventMask, &wa);
 		XDefineCursor(dpy, m->barwin, cursor[CurNormal]->cursor);
 		XMapRaised(dpy, m->barwin);
+		XSetClassHint(dpy, m->barwin, ch);
 	}
+	XFree(ch);
 }
 
 void
-- 
2.15.0

