From a48c499b53f6f8050319edc2f31c8f00acdaa50a Mon Sep 17 00:00:00 2001
From: Omar Sandoval <osandov@osandov.com>
Date: Sun, 10 Jan 2016 01:50:22 -0800
Subject: [PATCH] Set class name on status bar

This is useful for, e.g., compton.
---
 dwm.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dwm.c b/dwm.c
index ff7e096f61fc..21a2684874da 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1806,6 +1806,9 @@ updatebars(void)
 		.background_pixmap = ParentRelative,
 		.event_mask = ButtonPressMask|ExposureMask
 	};
+	XClassHint *ch = XAllocClassHint();
+	ch->res_name = "dwmstatus";
+	ch->res_class = "dwm";
 	for (m = mons; m; m = m->next) {
 		if (m->barwin)
 			continue;
@@ -1814,7 +1817,9 @@ updatebars(void)
 		                          CWOverrideRedirect|CWBackPixmap|CWEventMask, &wa);
 		XDefineCursor(dpy, m->barwin, cursor[CurNormal]->cursor);
 		XMapRaised(dpy, m->barwin);
+		XSetClassHint(dpy, m->barwin, ch);
 	}
+	XFree(ch);
 }
 
 void
-- 
2.7.0

