From f2cebaf5c74f9cd989916f8fe8d39127b2ceb20f Mon Sep 17 00:00:00 2001
From: Omar Sandoval <osandov@osandov.com>
Date: Sat, 23 Jan 2016 15:07:21 -0800
Subject: [PATCH] Set the EWMH _NET_SUPPORTING_WM_CHECK atom

---
 dwm.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/dwm.c b/dwm.c
index df8246250c29..95a5f036601c 100644
--- a/dwm.c
+++ b/dwm.c
@@ -63,7 +63,7 @@
 /* enums */
 enum { CurNormal, CurResize, CurMove, CurLast }; /* cursor */
 enum { SchemeNorm, SchemeSel, SchemeLast }; /* color schemes */
-enum { NetSupported, NetWMName, NetWMState,
+enum { NetSupported, NetSupportingWMCheck, NetWMName, NetWMState,
        NetWMFullscreen, NetActiveWindow, NetWMWindowType,
        NetWMWindowTypeDialog, NetClientList, NetLast }; /* EWMH atoms */
 enum { WMProtocols, WMDelete, WMState, WMTakeFocus, WMLast }; /* default atoms */
@@ -1609,6 +1609,7 @@ setup(void)
 	wmatom[WMTakeFocus] = XInternAtom(dpy, "WM_TAKE_FOCUS", False);
 	netatom[NetActiveWindow] = XInternAtom(dpy, "_NET_ACTIVE_WINDOW", False);
 	netatom[NetSupported] = XInternAtom(dpy, "_NET_SUPPORTED", False);
+	netatom[NetSupportingWMCheck] = XInternAtom(dpy, "_NET_SUPPORTING_WM_CHECK", False);
 	netatom[NetWMName] = XInternAtom(dpy, "_NET_WM_NAME", False);
 	netatom[NetWMState] = XInternAtom(dpy, "_NET_WM_STATE", False);
 	netatom[NetWMFullscreen] = XInternAtom(dpy, "_NET_WM_STATE_FULLSCREEN", False);
@@ -1842,8 +1843,10 @@ updatebars(void)
 		.event_mask = ButtonPressMask|ExposureMask
 	};
 	XClassHint *ch = XAllocClassHint();
+	Atom utf8string;
 	ch->res_name = "dwmstatus";
 	ch->res_class = "dwm";
+	utf8string = XInternAtom(dpy, "UTF8_STRING", False);
 	for (m = mons; m; m = m->next) {
 		if (m->barwin)
 			continue;
@@ -1853,7 +1856,14 @@ updatebars(void)
 		XDefineCursor(dpy, m->barwin, cursor[CurNormal]->cursor);
 		XMapRaised(dpy, m->barwin);
 		XSetClassHint(dpy, m->barwin, ch);
+		XChangeProperty(dpy, m->barwin, netatom[NetWMName], utf8string,
+				8, PropModeReplace, (unsigned char *)"dwm", 3);
 	}
+	/* it doesn't matter which bar we use for this, so just use the first one */
+	XChangeProperty(dpy, root, netatom[NetSupportingWMCheck], XA_WINDOW, 32,
+			PropModeReplace, (unsigned char *)&mons->barwin, 1);
+	XChangeProperty(dpy, mons->barwin, netatom[NetSupportingWMCheck], XA_WINDOW,
+			32, PropModeReplace, (unsigned char *)&mons->barwin, 1);
 	XFree(ch);
 }
 
-- 
2.7.0

